SocketExtensions/AsyncSocket.cs

```diff
@@ -33,9 +46,12 @@ public AsyncSocket(Socket s)
             return new ReadResult(bytesRead, buffer);
         }
 
-        public async Task SendMessage(byte[] buffer)
+        public async void SendMessage(byte[] buffer)
         {
+            _sender = new SocketAwaitableEventWrapper();
```

Comment:
this is very shitty for performance. instead we should use a consumer/producer pattern